src/context/Organization.ts

```diff
@@ -1,45 +1,66 @@
  */
 export class Organization implements IOrganization {
-    private readonly attributes: IOrganizationAttributes;
-
-    public readonly baseCurrencyId: string;
-    public readonly defaultCountryCode: string;
-    public readonly isAutoSaveEnabled: boolean;
-    public readonly languageId: number;
-    public readonly organizationId: string;
-    public readonly uniqueName: string;
-    public readonly useSkypeProtocol: boolean;
-
-    constructor(public readonly clientUrl: string, settings: Xrm.OrganizationSettings, private readonly discoveryUrl?: string) {
-        Object.assign(this, settings);
-        this.attributes = (settings as any).attributes || {};
+    private readonly _clientUrl: Url;
+
+    constructor(clientUrl: string, private readonly _settings: Xrm.OrganizationSettings) {
+        if (clientUrl) {
+            this._clientUrl = new Url(clientUrl);
+        } else {
+            const locationUrl: Url = new Url(location.href);
+            this._clientUrl = new Url(locationUrl.origin);
+        }
     }
 
-    get name(): string { return this.attributes.name; }
+    get baseCurrencyId(): string { return this._settings.baseCurrencyId; }
+    get defaultCountryCode(): string { return this._settings.defaultCountryCode; }
+    get languageId(): number { return this._settings.languageId; }
+    get organizationId(): string { return this._settings.organizationId; }
+    get uniqueName(): string { return this._settings.uniqueName; }
+    get isAutoSaveEnabled(): boolean { return this._settings.isAutoSaveEnabled; }
+    get useSkypeProtocol(): boolean { return this._settings.useSkypeProtocol; }
+
+    getClientUrl(): string {
+        return this._clientUrl.toString();
+    }
+
+    getDiscoveryOriginUrl(): string {
+        const discoUrl: string = this.getDiscoveryUrl();
+        return discoUrl ? new Url(discoUrl).origin : undefined;
+    }
+
+    /**
+     * Returns the discovery url associated to the organization.
+     * @see {@link https://docs.microsoft.com/en-us/dynamics365/customer-engagement/developer/org-service/discover-url-organization-organization-service }.
+     */
+    getDiscoveryUrl(): string {
+        const subpath: string = this.getClientUrlHostnameSubpath();
+        return subpath ? `https://disco.${subpath}/XRMServices/2011/Discovery.svc` : undefined;
+    }
 
-    public getDiscoveryUrl(): string {
-        return this.discoveryUrl || this.getDefaultDiscoveryUrl();
+    getName(): string {
+        const subpath: string = this.getClientUrlHostnameSubpath();
+        return subpath
+            ? this._clientUrl.hostname.replace(`.${subpath}`, "")
+            : undefined;
     }
 
-    private getDefaultDiscoveryUrl(): string {
-        return this.clientUrl && this.name
-            ? this.clientUrl.replace(this.name, discoveryUrlPrefix)
-            : this.clientUrl;
+    private getClientUrlHostnameSubpath(): string {
```

Comment:
This method could be renamed to `getRegionHostName`. I think it would fit better with [Microsoft terminology](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/developer/developer-resources-page).

You may also use a regular expression to capture the part of the host name that represents the region.

`/\.(crm\d*\.dynamics\.com)$/`